### PR TITLE
[CIT-309] thinedge crate

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
Created a separate thin edge crate. The tests are still part of the c8y translator code, maybe need to add a new set of tests to this library.

### CODE REVIEW

**Review by mneumann — COMMENTED** [review]
...

File: mapper/tedge_json/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+
+pub struct MultiValueMeasurement {
+    pub name: String,
+    pub values: Vec<SingleValueMeasurement>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct CumulocityJson {
+    c8y_json: JsonValue,
+}
```
Comment (didier-wenzek) [inline mapper/tedge_json/src/lib.rs:35]: This is unrelated to `ThinEdgeJson` and MUST stay in `c8y_translator_lib`.

File: mapper/tedge_json/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+        } else {
+            Ok(MultiValueMeasurement {
+                name: String::from(name),
+                values: single_values,
+            })
+        }
+    }
+}
+
+pub fn input_prefix(input: &str, len: usize) -> String {
```
Comment (didier-wenzek) [inline mapper/tedge_json/src/lib.rs:180]: This is an helper function and should not be `pub`.

=> The tests have to be moved from `c8y_translator_lib/lib.rs` to here, not that function made `pub` to please Cargo.

File: mapper/cumulocity/c8y_translator_lib/src/lib.rs [hunk 3]
```diff
@@ -820,22 +549,22 @@ mod tests {
     fn prefix_fn_removes_extra_chars() {
         let input = "薄いエッジ";
         assert_eq!(input.len(), 15);
-        assert_eq!(input_prefix(input, 1), "薄");
+        assert_eq!(tedge_json::input_prefix(input, 1), "薄");
     }
```
Comment (didier-wenzek) [inline mapper/cumulocity/c8y_translator_lib/src/lib.rs:553]: This test of the private helper function should be moved where the function is defined.

File: mapper/cumulocity/c8y_translator_lib/src/lib.rs [hunk 4]
```diff
@@ -820,22 +549,22 @@ mod tests {
...
 
     #[test]
     fn prefix_fn_let_unchanged_short_inputs() {
         let input = "FØØ";
         assert_eq!(input.len(), 5);
-        assert_eq!(input_prefix(input, 4), input);
+        assert_eq!(tedge_json::input_prefix(input, 4), input);
```
Comment (didier-wenzek) [inline mapper/cumulocity/c8y_translator_lib/src/lib.rs:559]: This test of the private helper function should be moved where the function is defined.

File: mapper/cumulocity/c8y_translator_lib/src/lib.rs [hunk 5]
```diff
@@ -820,22 +549,22 @@ mod tests {
...
     }
 
     use proptest::prelude::*;
 
     proptest! {
         #[test]
         fn prefix_doesnt_crash(input in "\\PC*") {
-            let _ = input_prefix(&input, 10);
+            let _ = tedge_json::input_prefix(&input, 10);
```
Comment (didier-wenzek) [inline mapper/cumulocity/c8y_translator_lib/src/lib.rs:567]: This test of the private helper function should be moved where the function is defined.

File: mapper/tedge_json/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+        match input {
+            JsonValue::String(_) | JsonValue::Short(_) => "a string",
+            JsonValue::Number(_) => "a number",
+            JsonValue::Object(_) => "an object",
+            JsonValue::Array(_) => "an array",
+            JsonValue::Boolean(_) => "a boolean",
+            JsonValue::Null => "null",
+        }
+    }
+}
```
Comment (didier-wenzek) [inline mapper/tedge_json/src/lib.rs:287]: Missing new line and the end of the file.

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
The new crate must contain only ThinEdgeJson related code and not a single line related to CumulocityJson. And vice versa, the former crate `c8y_translator_lib` must contain only Cumulocity related code.


* In particular, all the tests that are not c8y specific as `thin_edge_json_reject_invalid_utf8` must be moved in the new thin-edge-json crate. These tests are expecting an error from `CumulocityJson::from_thin_edge_json()` but must get the very same expected error from `ThinEdgeJson::read_from_bytes()`.
* A better name for this crate might be `thin_edge_json` since this is the public name for the format.

File: mapper/tedge_json/src/lib.rs (hunk 1 above)
Comment (PradeepKiruvale) [reply mapper/tedge_json/src/lib.rs:35]: yes, moved.

File: mapper/tedge_json/src/lib.rs (hunk 2 above)
Comment (PradeepKiruvale) [reply mapper/tedge_json/src/lib.rs:180]: fixed

File: mapper/cumulocity/c8y_translator_lib/src/lib.rs (hunk 3 above)
Comment (PradeepKiruvale) [reply mapper/cumulocity/c8y_translator_lib/src/lib.rs:553]: fixed

File: mapper/cumulocity/c8y_translator_lib/src/lib.rs (hunk 4 above)
Comment (PradeepKiruvale) [reply mapper/cumulocity/c8y_translator_lib/src/lib.rs:559]: fixed

File: mapper/cumulocity/c8y_translator_lib/src/lib.rs (hunk 5 above)
Comment (PradeepKiruvale) [reply mapper/cumulocity/c8y_translator_lib/src/lib.rs:567]: fixed

File: mapper/tedge_json/src/lib.rs (hunk 6 above)
Comment (PradeepKiruvale) [reply mapper/tedge_json/src/lib.rs:287]: fixed

File: mapper/thin_edge_json/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,574 @@
...
+#[derive(Serialize, Deserialize)]
+pub struct MultiValueMeasurement {
+    pub name: String,
+    pub values: Vec<SingleValueMeasurement>,
+}
+
+impl ThinEdgeJson {
+    pub fn read_from_bytes(input: &[u8]) -> Result<Vec<u8>, ThinEdgeJsonError> {
+        Ok(ThinEdgeJson::from_utf8(input, Utc::now())?.serialize_thin_edge())
+    }
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:40]: What is the purpose of this function?

* At least badly named. When I see `ThinEdgeJson::read_form_bytes()`, I expect a function taking some bytes and returning a `ThinEdgeJson` struct not a different sequence of bytes.
* Seems to be used only for tests expecting an error. These tests can be re-written by directly calling `ThinEdgeJson::from_utf8()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,574 @@
...
+                })?;
+                Ok(String::from(timestamp))
+            }
+            _ => Err(ThinEdgeJsonError::new_invalid_json_time(value)),
+        }
+    }
+
+    pub fn serialize_thin_edge(&self) -> Vec<u8> {
+        serde_json::to_vec(self).unwrap()
+    }
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:132]: What is the layout of the JSON output? Unlikely some ThinEdgeJson.

* This function has to be removed or fixed.

File: mapper/thin_edge_json/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,574 @@
...
+
+    pub fn deserialize_thin_edge(data: Vec<u8>) -> Self {
+        serde_json::from_slice(&data).unwrap()
+    }
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:136]: What is the expected JSON input layout? Unlikely some ThinEdgeJson.

* This function has to be removed or fixed to replace `ThinEdgeJson::from_utf8()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,574 @@
...
+mod tests {
+    use super::*;
+
+    #[test]
+    fn thin_edge_json_reject_invalid_utf8() {
+        let input = b"temperature\xc3\x28";
+
+        let expected_error =
+            r#"Invalid UTF8: invalid utf-8 sequence of 1 bytes from index 11: temperature..."#;
+        let output = ThinEdgeJson::read_from_bytes(input);
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:312]: This test should use `ThinEdgeJson::from_utf8()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,574 @@
...
+        let error = output.unwrap_err();
+        assert_eq!(expected_error, error.to_string());
+    }
+
+    #[test]
+    fn thin_edge_json_reject_non_utf8_input() {
+        let input = b"\xc3\x28";
+
+        let expected_error = r#"Invalid UTF8: invalid utf-8 sequence of 1 bytes from index 0: ..."#;
+        let output = ThinEdgeJson::read_from_bytes(input);
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:323]: This test should use `ThinEdgeJson::from_utf8()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,574 @@
...
+        let error = output.unwrap_err();
+        assert_eq!(expected_error, error.to_string());
+    }
+
+    #[test]
+    fn thin_edge_json_reject_arrays() {
+        let input = r"[50,23]";
+
+        let expected_error = r#"Invalid Thin Edge measurement: it cannot be an array: [50,23]"#;
+        let output = ThinEdgeJson::read_from_bytes(&String::from(input).into_bytes());
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:334]: This test should use `ThinEdgeJson::from_str()`.

1. You expect an error, so you don't need this weird `ThinEdgeJson::read_from_bytes` function.
2. You have a string as input and a function to read a string. There is no point to build a temp `[u8]`.

File: mapper/thin_edge_json/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,574 @@
...
+    #[test]
+    fn thin_edge_json_reject_nested_arrays() {
+        let input = r#"{
+           "time" : "2013-06-22T17:03:14.000+02:00",
+           "temperature": [50,60,70]
+          }"#;
+
+        let expected_error =
+            r#"Not a number: the "temperature" value must be a number, not an array."#;
+        let output = ThinEdgeJson::read_from_bytes(&String::from(input).into_bytes());
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:349]: This test should use `ThinEdgeJson::from_str()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,574 @@
...
+    fn thin_edge_json_reject_string_value() {
+        let input = r#"{
+           "time" : "2013-06-22T17:03:14.000+02:00",
+           "temperature": 50,
+           "pressure": "20"
+          }"#;
+
+        let expected_error =
+            r#"Not a number: the "pressure" value must be a number, not a string."#;
+        let output = ThinEdgeJson::read_from_bytes(&String::from(input).into_bytes());
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:365]: This test should use `ThinEdgeJson::from_str()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,574 @@
...
+
+    #[test]
+    fn thin_edge_json_reject_huge_number() {
+        let input = r#"{
+           "time" : "2013-06-22T17:03:14.000+02:00",
+           "temperature": 10e99999
+          }"#;
+
+        let expected_error = r#"Number out-of-range: the "temperature" value is too large to be represented as a float64."#;
+        let output = ThinEdgeJson::read_from_bytes(&String::from(input).into_bytes());
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:379]: This test should use `ThinEdgeJson::from_str()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,574 @@
...
+        let string_value_thin_edge_json = r#"{
+           "time" : "2013-06-22T17:03:14.000+02:00",
+           "temperature": true,
+           "pressure": 220
+          }"#;
+
+        let expected_output =
+            r#"Not a number: the "temperature" value must be a number, not a boolean."#;
+        let output =
+            ThinEdgeJson::read_from_bytes(&String::from(string_value_thin_edge_json).into_bytes());
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:396]: This test should use `ThinEdgeJson::from_str()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,574 @@
...
+                         "breadth": 32.54,
+                         "depth": 117.67
+                      }
+                  },
+                "pressure": 98
+        }"#;
+        let expected_output =
+            r#"More than 2 nested levels: the record for "area" must be flattened."#;
+        let output =
+            ThinEdgeJson::read_from_bytes(&String::from(multi_level_heirarchy).into_bytes());
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:419]: This test should use `ThinEdgeJson::from_str()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,574 @@
...
+    fn thin_edge_reject_measurement_named_type() {
+        let string_value_thin_edge_json = r#"{
+           "time" : "2013-06-22T17:03:14.000+02:00",
+           "type": 40,
+           "pressure": 220
+          }"#;
+
+        let expected_output = r#"Invalid measurement name: "type" is a reserved word."#;
+        let output =
+            ThinEdgeJson::read_from_bytes(&String::from(string_value_thin_edge_json).into_bytes());
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:434]: This test should use `ThinEdgeJson::from_str()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,574 @@
...
+
+    #[test]
+    fn thin_edge_json_reject_invalid_json() {
+        let input = r#"{
+           "time" : "2013-06-22T17:03:14.000+02:00",
+           "pressure": 220;
+          }"#;
+
+        let expected_error = r#"Invalid JSON: Unexpected character: ; at (3:27): {"time":"2013-06-22T17:03:14.000+02:00","pressure":220;}"#;
+        let output = ThinEdgeJson::read_from_bytes(&String::from(input).into_bytes());
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:463]: This test should use `ThinEdgeJson::from_str()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,574 @@
...
+
+    #[test]
+    fn thin_edge_json_reject_partial_json() {
+        let input = r#"{
+           "time" : "2013-06-22T17:03:14.000+02:00",
+        "#;
+
+        let expected_error =
+            r#"Invalid JSON: Unexpected end of JSON: {"time":"2013-06-22T17:03:14.000+02:00","#;
+        let output = ThinEdgeJson::read_from_bytes(&String::from(input).into_bytes());
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:477]: This test should use `ThinEdgeJson::from_str()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,574 @@
...
+        assert_eq!(expected_error, error.to_string());
+    }
+
+    #[test]
+    fn thin_edge_json_reject_empty_record() {
+        let input = "{}";
+
+        let expected_error =
+            "Empty Thin Edge measurement: it must contain at least one measurement";
+        let output = ThinEdgeJson::read_from_bytes(&String::from(input).into_bytes());
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:489]: This test should use `ThinEdgeJson::from_str()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,574 @@
...
+
+    #[test]
+    fn thin_edge_json_reject_just_time() {
+        let input = r#"{
+           "time" : "2013-06-22T17:03:14.000+02:00"
+        }"#;
+
+        let expected_error =
+            "Empty Thin Edge measurement: it must contain at least one measurement";
+        let output = ThinEdgeJson::read_from_bytes(&String::from(input).into_bytes());
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:503]: This test should use `ThinEdgeJson::from_str()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 23]
```diff
@@ -0,0 +1,574 @@
...
+
+    #[test]
+    fn thin_edge_json_reject_empty_measurement() {
+        let input = r#"{
+           "foo" : {}
+        }"#;
+
+        let expected_error =
+            r#"Empty Thin Edge measurement: "foo" must contain at least one measurement"#;
+        let output = ThinEdgeJson::read_from_bytes(&String::from(input).into_bytes());
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:517]: This test should use `ThinEdgeJson::from_str()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 24]
```diff
@@ -0,0 +1,574 @@
...
+
+    #[test]
+    fn thin_edge_json_reject_partial_timestamp() {
+        let input = r#"{
+           "time" : "2013-06-22",
+           "pressure": 220
+          }"#;
+
+        let expected_error = "Invalid ISO8601 timestamp (expected YYYY-MM-DDThh:mm:ss.sss.±hh:mm): \"2013-06-22\": premature end of input";
+        let output = ThinEdgeJson::read_from_bytes(&String::from(input).into_bytes());
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:531]: This test should use `ThinEdgeJson::from_str()`.

File: mapper/thin_edge_json/src/lib.rs [hunk 25]
```diff
@@ -0,0 +1,574 @@
...
+    #[test]
+    fn thin_edge_json_reject_invalid_timestamp() {
+        let input = r#"{
+           "time" : "2013-06-22 3am",
+           "pressure": 220
+          }"#;
+
+        let expected_error =
+            "Invalid ISO8601 timestamp (expected YYYY-MM-DDThh:mm:ss.sss.±hh:mm): \"2013-06-22 3am\": input contains invalid characters";
+        let output = ThinEdgeJson::read_from_bytes(&String::from(input).into_bytes());
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/lib.rs:546]: This test should use `ThinEdgeJson::from_str()`.

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
* The function `ThinEdgeJson::read_form_bytes()` has to be removed.
* The functions `ThinEdgeJson::serialize_thin_edge()` and ThinEdgeJson::deserialize_thin_edge()`  are not emitting/reading ThinEdgeJson content. They have to removed from this PR and implemented properly later.

**Review by didier-wenzek — APPROVED** [review]
Approved as a first step of [CIT-309] - i.e. decoupling ThinEdgeJson from Cumulocity Json.

Follow-up tasks:
* Add tests on the happy path of `ThinEdgeJson::from_utf8()`. The current tests only check for errors.
* Change the type of timestamp to `DateTime<Utc>`.
* Implement the serialisation symetric to `ThinEdgeJson::from_utf8()`.

**Review by mneumann — APPROVED** [review]
Very well done.

**Comment by PradeepKiruvale** [comment]
This PR just segregates the thin edge json code from c8y json code. Now both thin edge json and c8y json are different crates.
